src/main/java/de/tum/in/www1/artemis/service/QuizService.java

```diff
@@ -0,0 +1,295 @@
+package de.tum.in.www1.artemis.service;
+
+import java.util.*;
+import java.util.function.Function;
+
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.quiz.*;
+import de.tum.in.www1.artemis.repository.DragAndDropMappingRepository;
+import de.tum.in.www1.artemis.repository.ShortAnswerMappingRepository;
+
+@Service
+public abstract class QuizService<T extends QuizConfiguration> {
+
+    private final DragAndDropMappingRepository dragAndDropMappingRepository;
+
+    private final ShortAnswerMappingRepository shortAnswerMappingRepository;
+
+    /**
+     * Save the given QuizConfiguration to the database according to the implementor.
+     *
+     * @param quizConfiguration the QuizConfiguration to be saved.
+     * @return the saved QuizConfiguration
+     */
+    protected abstract T saveAndFlush(T quizConfiguration);
+
+    protected QuizService(DragAndDropMappingRepository dragAndDropMappingRepository, ShortAnswerMappingRepository shortAnswerMappingRepository) {
+        this.dragAndDropMappingRepository = dragAndDropMappingRepository;
+        this.shortAnswerMappingRepository = shortAnswerMappingRepository;
+    }
+
+    /**
+     * Save the given QuizConfiguration
+     *
+     * @param quizConfiguration the QuizConfiguration to be saved
+     * @return saved QuizConfiguration
+     */
+    public T save(T quizConfiguration) {
+        // fix references in all questions (step 1/2)
+        for (var quizQuestion : quizConfiguration.getQuizQuestions()) {
+            if (quizQuestion instanceof MultipleChoiceQuestion mcQuestion) {
+                fixReferenceMultipleChoice(mcQuestion);
+            }
+            else if (quizQuestion instanceof DragAndDropQuestion dndQuestion) {
+                fixReferenceDragAndDrop(dndQuestion);
+            }
+            else if (quizQuestion instanceof ShortAnswerQuestion saQuestion) {
+                fixReferenceShortAnswer(saQuestion);
+            }
+        }
+
+        T savedQuizConfiguration = saveAndFlush(quizConfiguration);
+
+        // fix references in all drag and drop questions and short answer questions (step 2/2)
+        for (QuizQuestion quizQuestion : savedQuizConfiguration.getQuizQuestions()) {
+            if (quizQuestion instanceof DragAndDropQuestion dragAndDropQuestion) {
+                // restore references from index after save
+                restoreCorrectMappingsFromIndicesDragAndDrop(dragAndDropQuestion);
+            }
+            else if (quizQuestion instanceof ShortAnswerQuestion shortAnswerQuestion) {
+                // restore references from index after save
+                restoreCorrectMappingsFromIndicesShortAnswer(shortAnswerQuestion);
+            }
+        }
+
+        return savedQuizConfiguration;
+    }
+
+    private void fixReferenceMultipleChoice(MultipleChoiceQuestion mcQuestion) {
```

Comment:
This is the (almost identical) same function three times in a row.

Reply:
refactored